### PR TITLE
New environment structure

### PR DESCRIPTION
**Pull request by hvater**

### CODE REVIEW

File: examples/example_gymwrapper.ipynb
Comment (hvater) [inline examples/example_gymwrapper.ipynb:1]: Please make sure that all notebook outputs are deleted before the notebook is commited.

Do that either manually or use something like this: https://github.com/kynan/nbstripout

File: exciting_environments/classic_core_env.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        # )
+
+        # generate terminated
+
+        # get last state so that the simulation can be continued from the end point
+        last_state = tree_unflatten(struct, jnp.array(states_flatten)[:, -1:])
+
+        # terminated = jax.vmap(self.generate_terminated, in_axes=(0, 0, self.in_axes_env_properties))(
+        #     states_without_init_state, reward, self.env_properties
+        # )
```
Comment (hvater) [inline exciting_environments/classic_core_env.py:221]: please delete stuff like this. Commit the changes with a sensible commit message. If we decide to roll it back, you can look for such a commit message and undo it

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 2]
```diff
@@ -50,8 +50,8 @@ def __init__(
         physical_constraints: dict = None,
         action_constraints: dict = None,
         static_params: dict = None,
+        control_state: list = None,
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:53]: I am not 100% sure if I understand what the goal is with control state. We should discuss this.

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 3]
```diff
@@ -125,7 +126,7 @@ class PhysicalState:
         omega: jax.Array
 
     @jdc.pytree_dataclass
-    class Optional:
+    class Additions:
         """Dataclass containing additional information for simulation."""
 
         something: jax.Array
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:132]: It looks kinda janky to have this, with `something` if it is set to `None` anyways

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 4]
```diff
@@ -276,37 +277,59 @@ def vector_field(t, y, args):
...
-            + 0.1 * (action[0] / action_constraints.force) ** 2
-        )
+        reward = 0
+        for name in self.control_state:
+            reward += -(
+                (
+                    (getattr(state.physical_state, name) - getattr(state.reference, name))
+                    / (getattr(env_properties.physical_constraints, name)).astype(float)
+                )
+                ** 2
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:325]: The angle information (angle of the pendulum) should be transformed to sin and cos here. Otherwise the loss is kind of inconsistent

File: exciting_environments/fluid_tank/fluid_tank_env.py
Comment (hvater) [inline exciting_environments/fluid_tank/fluid_tank_env.py:1]: Compare to comments on the cart pole:

- I would like to discuss the control state 
- The unused Addtions are somewhat questionable

File: exciting_environments/gym_wrapper.py [hunk 5]
```diff
@@ -37,55 +75,111 @@ def step(self, action):
...
+
+    def generate_new_ref(self, state, env_properties, hold_steps):
+        with jdc.copy_and_mutate(state, validate=False) as new_state:
+            init = self.env.init_state(env_properties, state.PRNGKey)
+            for name in self.control_state:
+                setattr(new_state.reference, name, getattr(init.physical_state, name))
+
+            hold_steps = jax.random.randint(
+                init.PRNGKey,
+                minval=self.ref_params["hold_steps_min"],
```
Comment (hvater) [inline exciting_environments/gym_wrapper.py:175]: i think i prefer these to be passed to the function and not be used through `self.` like this. Also is it a problem that this is a dict? Can this lead to inconsistent behaviour? Can you change this between resets?

File: exciting_environments/mass_spring_damper/mass_spring_damper_env.py
Comment (hvater) [inline exciting_environments/mass_spring_damper/mass_spring_damper_env.py:1]: Compare to comments on the cart pole:

- I would like to discuss the control state
- The unused Addtions are somewhat questionable

File: exciting_environments/pendulum/pendulum_env.py
Comment (hvater) [inline exciting_environments/pendulum/pendulum_env.py:1]: Compare to comments on the cart pole:

- I would like to discuss the control state
- The unused Addtions are somewhat questionable
- map angle information to sin and cos for the reward

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 6]
```diff
@@ -212,18 +211,30 @@ def vector_field(t, y, args):
...
 
-        phys = self.PhysicalState(
-            deflection=deflection_k1,
-            velocity=velocity_k1,
-            theta=theta_k1,
-            omega=omega_k1,
-        )
-        opt = None  # Optional(something=...)
-        return self.State(physical_state=phys, PRNGKey=None, optional=opt)
+        with jdc.copy_and_mutate(state, validate=False) as new_state:
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:214]: Should we investigate why `validate=False` is necessary here? I think it generally should not need to be, looking at the documentation for `jdc`

File: exciting_environments/core_env.py [hunk 7]
```diff
@@ -347,59 +146,53 @@ def vmap_sim_ahead(self, init_state, actions, obs_stepsize, action_stepsize):
...
     @partial(jax.jit, static_argnums=0)
-    @abstractmethod
-    def generate_terminated(self, state, reward):
-        """Returns terminated information."""
-        return
+    def vmap_init_state(self, rng: chex.PRNGKey = None):
+        """Returns default or random initial state for all batches."""
+        return jax.vmap(self.init_state, in_axes=(self.in_axes_env_properties, 0, 0))(
+            self.env_properties, rng, jnp.ones(self.batch_size)
+        )
```
Comment (hvater) [inline exciting_environments/core_env.py:177]: jit for vmap reset?

File: exciting_environments/cart_pole/cart_pole_env.py (hunk 6 above)
Comment (OliverSchw) [reply exciting_environments/cart_pole/cart_pole_env.py]: Seems to have caused problems in an older version. The validation is now activated again and works without any problems.

File: exciting_environments/core_env.py (hunk 7 above)
Comment (OliverSchw) [reply exciting_environments/core_env.py:178]: I have changed it.
